### DISCUSSION TITLE
Mixing AD kernels

### DISCUSSION THREAD

**Discussion by rtaylo45** [idaholab/moose#20075] [Q&A General] [answered]
Hey,

I was wondering if someone could tell me what happens if I mix AD and non-AD kernels? My guess is that the jacobian would not be correct and cause convergence problems. Would this also cause MOOSE to converge to a different solution? Is there anything else that would happen? Obviously this is not recommended, but im trying to understand what happens under the hood of MOOSE. 

Thanks,
Zack

**Comment by GiudGiud** [comment, marked as answer]
Hello

First, if you are **not** using Newton's method, ~~then nothing will change~~ . ~~(P)~~ JFNK does not make use of AD because we don't have to form a Jacobian.
EDIT: see discussion below for PJFNK

Then if you are using Newton's method,
If your non-AD kernels have the Jacobian (diagonal and off-diagonal) contributions coded exactly, then it will not change anything. The Jacobian will still be exact, and will still be used the same way.
If the Jacobian is only coded partially or approximately, then it may or may not cause convergence problems that would not have been there when using AD. There's really no certainty though.

docs on AD:
https://mooseframework.inl.gov/automatic_differentiation/#!

It'll not converge to a different solution though. The residual has not changed, and we're still converging to the (usually unique) solution that zeroes it out. 

Guillaume

**Reply by hugary1995** [reply]
@GiudGiud Is it accurate to say PJFNK doesn't use AD? I thought it still uses jacobian for preconditioning (well, it apparently depends on the specific preconditioner). I'm just curious.

**Reply by GiudGiud** [reply]
Umm that could be.
Let's tag @lindsayad to be sure

**Reply by GiudGiud** [reply]
We use the diagonal of the Jacobian for preconditioning

**Reply by hugary1995** [reply]
Yeah, that's the default I believe. If you use smp and specify `full = true` it'll use the entire Jacobian, in which case the incorrect Jacobian _might_ affect convergence.

**Reply by rtaylo45** [reply]
@hugary1995 So if we change the default preconditioner for PJFNK to one what doesn't use the jacobian then we should be fine right?

**Reply by hugary1995** [reply]
That's JFNK then. It depends on what you mean by "should be fine". It would be fine in the sense that mixing AD and nonAD kernels makes no difference. But I've had limited success with JFNK in practice.

**Reply by lindsayad** [reply]
Jacobian-free methods typically require preconditioning in order to be efficient. In theory you could converge to a solution in n-dof GMRES iterations without any preconditioning but there are some caveats there in practice I believe.

@hugary1995 whether a kernel is AD or not doesn't make a difference with respect to whether a `computeJacobian` function will get called. If you have any solve method that constructs a matrix (NEWTON or PJFNK for example) then `computeJacobian` methods will get called for both non-AD and AD kernels. You really should not think of ADKernels and Kernels as being any different with the exception that you can be quite confident that the former will produce a 100% accurate Jacobian while the latter may or may not. If the latter is 100% accurate, then with respect to the solver the ADKernel and Kernel will appear completely equivalent.

There is *one* tricky thing that AD triggers for PJFNK. For the default AD container type, we compute the object Jacobian all at once and we add to the system Jacobian all at once. It is impossible (or it would be very difficult and slow) to filter out Jacobian contributions from specific variables. Consequently if you have any AD objects and you are using our default MOOSE configuration, you will be implicitly getting a `full = true` matrix all the time, e.g. if you requested only a diagonal matrix or if you asked for specific row-column coupling, we would ignore your request. 

Perhaps I should also note that if you specify `solve_type = NEWTON` then we also set `full = true` behind the scenes and this is true irrespective of whether there are any AD objects in the simulation. We do this because it only makes sense to use NEWTON if you are supplying all the variable couplings (e.g. complete and hopefully accurate Jacobian). A differenc here is that I believe we still honor requests for specific row-column couplings ... but it would be to your detriment.

**Reply by hugary1995** [reply]
> You really should not think of ADKernels and Kernels as being any different with the exception that you can be quite confident that the former will produce a 100% accurate Jacobian while the latter may or may not. If the latter is 100% accurate, then with respect to the solver the ADKernel and Kernel will appear completely equivalent.

The op was assuming incorrect non-AD Jacobian.

I agree with you otherwise.
